source/fab/steps/compile_fortran.py

```diff
@@ -0,0 +1,129 @@
+            if errors:
+                # todo: do we need to fail if we're building a library? might not need that broken file...
+                err_str = "\n\n".join(map(str, errors))
+                raise RuntimeError(f"Error in compiling pass: {err_str}")
+
+            # check what we did compile
+            compiled_this_pass: Set[CompiledFile] = set(by_type(results_this_pass, CompiledFile))
+            per_pass.append(len(compiled_this_pass))
+            if len(compiled_this_pass) == 0:
+                logger.error("nothing compiled this pass")
+                break
+
+            # remove compiled files from list
+            logger.debug(f"compiled {len(compiled_this_pass)} files")
+
+            # results are not the same instances as passed in, due to mp copying
+            compiled_fpaths = {i.analysed_file.fpath for i in compiled_this_pass}
+            all_compiled.extend(compiled_this_pass)
+            already_compiled_files.update(compiled_fpaths)
+
+            # remove from remaining to compile
+            to_compile = set(filter(lambda af: af.fpath not in compiled_fpaths, to_compile))
+
+        log_or_dot_finish(logger)
+        logger.debug(f"compiled per pass {per_pass}")
+        logger.info(f"total fortran compiled {sum(per_pass)}")
+
+        if to_compile:
+            logger.debug(f"there were still {len(to_compile)} files left to compile")
+            for af in to_compile:
+                logger.debug(af.fpath)
+            logger.error(f"there were still {len(to_compile)} files left to compile")
+            exit(1)
+
+        artefacts['compiled_fortran'] = all_compiled
+
+    def get_compile_next(self, already_compiled_files: Set[Path], to_compile: Set[AnalysedFile]):
+
+        # find what to compile next
+        compile_next = set()
+        not_ready: Dict[Path, List[Path]] = {}
+        for af in to_compile:
+            # all deps ready?
+            unfulfilled = [dep for dep in af.file_deps if dep not in already_compiled_files and dep.suffix == '.f90']
+            if not unfulfilled:
+                compile_next.add(af)
+            else:
+                not_ready[af.fpath] = unfulfilled
+
+        # unable to compile anything?
+        if len(to_compile) and not compile_next:
+            all_unfulfilled: Set[Path] = set()
+            for unfulfilled in not_ready.values():
+                all_unfulfilled = all_unfulfilled.union(unfulfilled)
+            raise RuntimeError(
+                f"Nothing more can be compiled due to unfulfilled dependencies: {', '.join(map(str, all_unfulfilled))}")
+
+        return compile_next
+
+    def compile_file(self, analysed_file: AnalysedFile):
+        command = [self.exe]
+        command.extend(self.flags.flags_for_path(analysed_file.fpath, self._config.workspace))
+        command.append(str(analysed_file.fpath))
+
+        output_fpath = analysed_file.fpath.with_suffix('.o')
```

Comment:
It's handled automatically by the compiler with the [`-J` flag](https://github.com/metomi/fab/pull/100/files/0774139063cbb0e8b50aa95cb5ddfb711c4f5f87#diff-5965519ebf50cd663601d7c84a6d78d76a6c749bb0fe85ec43c7b486ff90cff2R184).

Can't the .o file be used to inform incremental builds just as well as the .mod file?

Reply:
No it can't. The `.mod` file represents the API of the module while the `.o` file represents the implementation. Just because the source has changed and therefore the implementation it doesn't mean the API has changed. Thus building an object file depends on the module files, not other object files. If the API hasn't changed then things which use that API don't need to be rebuilt. Only relinked which is always a universal process at the end. i.e. there is no such thing as an incremental link.